db/snapshot/backup_task.cc

```diff
@@ -147,6 +148,17 @@ future<> backup_task_impl::do_backup() {
                 auto [desc, ks, cf] = sstables::parse_path(file_path);
                 sstable_comps[desc.generation].emplace_back(name);
                 ++num_sstable_comps;
+
+                if (desc.component == sstables::component_type::Data) {
+                    auto st = co_await file_stat(file_path.native());
+                    // If the sstable is already unlinked after the snapshot was taken
+                    // track its generation in the unlinked_sstables list
+                    // so it can be prioritized for backup
+                    if (st.number_of_links == 1) {
```

Comment:
That's not very accurate. If an sstable is snapshot-ed twice (or more), then removed, this check would be false

Reply:
A deterministic way to verify it is to build an inode set from data file of active sstables, and then see if inode of data files in backup is present in that set. It's very file system specific, but so is this approach, and we can hide it behind an interface. Maybe there's a simpler way?

Reply:
In any case, if an sstable is indeed hardlink to multiple snapshots it wouldn't help to upload it earlier since unlinking it from this snapshot won't free up its capacity.
Not sire how imporant ot is to handle this corner case.

Reply:
so to some extent your patch makes sense, since you will be priorizing inodes that can be wiped after we're done with the uploading. it makes sense to pick the sstable with highest chance of freeing up space, if we're in a danger usage zone.

Reply:
> so to some extent your patch makes sense, since you will be priorizing inodes that can be wiped after we're done with the uploading. it makes sense to pick the sstable with highest chance of freeing up space, if we're in a danger usage zone.

I believe we should do it any time, even before getting into the danger zone, to reduce scale out events, otherwise I'm afraid that every major backup at ose to 90% utilization might trigger scale out.  It may work but it's expensive.

Reply:
you're right. please elaborate on what's expensive. retrieving link count is just about copying memory from kernel to user space.

Reply:
No no, what is expensive is launching new instances and migrating tablets back and forth assuming that when backup completes, scale-in would be desired as utilization would drop back below the low water mark.

Reply:
stat'ing each file when listing the snapshot directory is definitely not free, but it shouldn't be more expensive than regular random reads.